src/formula/variant_private.cpp

```diff
@@ -0,0 +1,282 @@
+{
+	return value_cast<variant_callable>(other)->callable_->less(callable_.get());
+}
+
+std::string variant_string::get_serialized_string() const
+{
+	std::ostringstream ss;
+	ss << "'";
+
+	for(const auto& c : string_) {
+		switch(c) {
+		case '\'':
+			ss << "[']";
+			break;
+		case '[':
+			ss << "[(]";
+			break;
+		case ']':
+			ss << "[)]";
+			break;
+		default:
+			ss << c;
+			break;
+		}
+	}
+
+	ss << "'";
+
+	return ss.str();
+}
+
+template<typename T>
+std::string variant_container<T>::to_string_impl(bool annotate, bool annotate_empty, mod_func_t mod_func) const
+{
+	std::ostringstream ss;
+
+	if(annotate) {
+		ss << "[";
+	}
+
+	bool first_time = true;
+
+	for(const auto& member : container_) {
+		if(!first_time) {
+			ss << ",";
+		}
+
+		first_time = false;
+
+		ss << to_string_detail(member, mod_func);
+	}
+
+	// TODO: evaluate if this really needs to be separately conditional.
+	if(annotate_empty && container_.empty()) {
+		ss << "->";
+	}
+
+	if(annotate) {
+		ss << "]";
+	}
+
+	return ss.str();
+}
+
+template<typename T>
+std::string variant_container<T>::string_cast() const
+{
+	return to_string_impl(false, false, [](const variant& v) { return v.string_cast(); });
+}
+
+template<typename T>
+std::string variant_container<T>::get_serialized_string() const
+{
+	return to_string_impl(true, true,   [](const variant& v) { return v.serialize_to_string(); });
+}
+
+template<typename T>
+std::string variant_container<T>::get_debug_string(const_formula_callable_vec* seen, bool verbose) const
+{
+	return to_string_impl(true, false, [&](const variant& v) { return v.to_debug_string(seen, verbose); });
+}
+
+// Force compilation of the following template instantiations
+template class variant_container<variant_vector>;
+template class variant_container<variant_map_raw>;
+
+variant variant_list::list_op(value_base_ptr second, std::function<variant(variant&, variant&)> op_func)
+{
+	const auto& other_list = value_cast<variant_list>(second);
+
+	if(num_elements() != other_list->num_elements()) {
+		throw type_error("List op requires two lists of the same length");
+	}
+
+	std::vector<variant> res;
+	res.reserve(num_elements());
+
+	for(size_t i = 0; i < num_elements(); ++i) {
+		res.push_back(op_func(get_container()[i], other_list->get_container()[i]));
+	}
+
+	return variant(&res);
+}
+
+bool variant_list::operator==(value_base_ptr other) const
+{
+	const auto& other_container = value_cast<variant_list>(other)->v_vector_;
+
+	if(num_elements() != other->num_elements()) {
+		return false;
+	}
+
+	for(size_t n = 0; n < num_elements(); ++n) {
+		if(v_vector_[n] != other_container[n]) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+bool variant_list::operator<=(value_base_ptr other) const
+{
+	const auto& other_container = value_cast<variant_list>(other)->v_vector_;
+
+	for(size_t n = 0; n != num_elements() && n != other->num_elements(); ++n) {
+		// TODO: shouldn't this just be <=?
```

Comment:
`==` is considered implicitly. If two elements are equal, it needs to check the next pair of elements to determine the result. Thus, the proper behaviour when they're equal is to continue to the next iteration of the loop.